test/model/message_test.dart

```diff
@@ -76,6 +94,335 @@ void main() {
+
+        async.elapse(kSendMessageOfferRestoreWaitPeriod - kLocalEchoDebounceDuration);
+        checkState().equals(OutboxMessageState.waitPeriodExpired);
+
+        store.takeOutboxMessage(store.outboxMessages.keys.single);
+        check(store.outboxMessages).isEmpty();
+      }));
+
+      test('failed -> (delete) because event received', () => awaitFakeAsync((async) async {
+        await prepareOutboxMessageToFailAfterDelay(Duration.zero);
+        await check(outboxMessageFailFuture).throws();
+        checkState().equals(OutboxMessageState.failed);
+
+        await receiveMessage();
+        check(store.outboxMessages).isEmpty();
+      }));
+
+      test('failed -> (delete) because outbox message was taken', () => awaitFakeAsync((async) async {
+        await prepareOutboxMessageToFailAfterDelay(Duration.zero);
+        await check(outboxMessageFailFuture).throws();
+        checkState().equals(OutboxMessageState.failed);
+
+        store.takeOutboxMessage(store.outboxMessages.keys.single);
+        check(store.outboxMessages).isEmpty();
+      }));
+    });
+
+    test('when sending to "(no topic)", process topic like the server does when creating outbox message', () => awaitFakeAsync((async) async {
+      await prepareOutboxMessage(
+        destination: StreamDestination(stream.streamId, TopicName('(no topic)')),
+        zulipFeatureLevel: 370);
+      async.elapse(kLocalEchoDebounceDuration);
+      check(store.outboxMessages).values.single
+        .conversation.isA<StreamConversation>().topic.equals(eg.t(''));
+    }));
+
+    test('legacy: when sending to "(no topic)", process topic like the server does when creating outbox message', () => awaitFakeAsync((async) async {
+      await prepareOutboxMessage(
+        destination: StreamDestination(stream.streamId, TopicName('(no topic)')),
+        zulipFeatureLevel: 369);
+      async.elapse(kLocalEchoDebounceDuration);
+      check(store.outboxMessages).values.single
+        .conversation.isA<StreamConversation>().topic.equals(eg.t('(no topic)'));
+    }));
+
+    test('set timestamp to now when creating outbox messages', () => awaitFakeAsync(
+      initialTime: eg.timeInPast,
+      (async) async {
+        await prepareOutboxMessage();
+        check(store.outboxMessages).values.single
+          .timestamp.equals(eg.utcTimestamp(eg.timeInPast));
+      },
+    ));
+  });
+
+  test('takeOutboxMessage', () async {
+    final stream = eg.stream();
+    await prepare(stream: stream);
+    await prepareMessages([]);
+
+    for (int i = 0; i < 10; i++) {
+      connection.prepare(apiException: eg.apiBadRequest());
+      await check(store.sendMessage(
+        destination: StreamDestination(stream.streamId, eg.t('topic')),
+        content: 'content')).throws();
+    }
+
+    final localMessageIds = store.outboxMessages.keys.toList();
+    store.takeOutboxMessage(localMessageIds.removeAt(5));
+    check(store.outboxMessages.keys).deepEquals(localMessageIds);
```

Comment:
nit: use checks-getter (like https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2101317278 )
```suggestion
    check(store.outboxMessages).keys.deepEquals(localMessageIds);
```